synapse/storage/user_directory.py

```diff
@@ -568,15 +591,14 @@ def search_user_dir(self, user_id, search_term, limit):
             where_clause = "1=1"
         else:
             join_clause = """
+                LEFT JOIN publicly_visible_users AS p USING (user_id)
                 LEFT JOIN (
-                    SELECT other_user_id AS user_id FROM users_who_share_public_rooms
-                    UNION
                     SELECT other_user_id AS user_id FROM users_who_share_private_rooms
```

Comment:
So, the problem here is that while `publicly_visible_users` returns only one row, `users_who_share_private_rooms` may now return multiple rows. Since this is a join that will lead to duplicates. Either, we need to do the same thing that we've done for `users_in_public_rooms` and have a separate table such that the join doesn't return multiple rows, OR we need to rearrange the query to correctly handle multiple rows.

I think we can use `EXISTS` to do that, which appears to give a sane ish query plan, i.e. something like:


```sql
SELECT d.user_id AS user_id, display_name, avatar_url
FROM user_directory_search AS t
INNER JOIN user_directory AS d USING (user_id)
WHERE
    vector @@ to_tsquery('english', ?)
    AND (
        EXISTS (select 1 from users_in_public_rooms WHERE user_id = t.user_id)
        OR EXISTS (
            SELECT 1 FROM users_who_share_private_rooms
            WHERE user_id = ? AND other_user_id = t.user_id
        ) 
    )
ORDER BY ...
LIMIT 5;
```

Reply:
Using that form also means we don't need to have a separate `publicly_visible_users` table

Reply:
I think the not-having-multiple-tables option is better. Done.